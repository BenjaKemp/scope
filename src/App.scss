
@import 'styles/_colors.scss'; 
@import 'styles/Dashboard.scss'; 
@import 'styles/Landing.scss'; 
@import 'styles/Player.scss'; 
@import 'styles/CenterPlayer.scss'; 

body{
	font-family: courier;
	color: black;
}

@font-face {
  font-family: mariokartfont;
  src: url('./fonts/@font-face/mariokartfont.woff2') format('woff2'),
       url('./fonts/@font-face/mariokartfont.woff') format('woff');
}

.container_wrapper {

    height: 50rem;
    overflow: auto;

}

.jumbotron_app {
	font-family: mariokartfont;
    background-color: $marioRed;
    height: 3rem;
    width: 100%;
    margin:0;
    padding:0;
    border: black;
    border-bottom: solid;

}




