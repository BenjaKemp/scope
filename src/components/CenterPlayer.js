import React from "react";
import { useState } from 'react';
import Jumbotron from 'react-bootstrap/Jumbotron';
import Container from 'react-bootstrap/Container';
import Modal from 'react-bootstrap/Modal';
import Button from 'react-bootstrap/Button';
import Table from 'react-bootstrap/Table';
import {useParams } from "react-router-dom";
import  Graph  from './Graphs';
import moment from 'moment';

const CenterPlayer = ({players, races}) => {
const [show, setShow] = useState(false);
const handleClose = () => setShow(false);
const handleShow = () => setShow(true);
const { id } = useParams();
const player  = players.find(el => el.id === id)
const pos  = players.indexOf(player)+1
const allPlayerRaces = races.filter(el => el.standings.includes(player.name) )

if(!player){
	return (
		<div></div>
		)
}
	return(
		<Jumbotron className ="container-full-bg center_player_background" fluid>
		  	<Container>
			   	<h2>Name: {player.name} - Current Position: {pos}</h2>
			    	<Table striped bordered hover>
						<thead>
							<tr>
								<th>Poition </th>
								<th>Frequency</th>
							</tr>
						</thead>
						<tbody>
							<tr>
								<td>1st</td>
								<td>{player.results[1]}</td>
							</tr>
							<tr>
								<td>2nd</td>
								<td>{player.results[2]}</td>
							</tr>
							<tr>
								<td>3rd</td>
								<td>{player.results[3]}</td>
							</tr>
							<tr>
								<td>4th</td>
								<td>{player.results[4]}</td>
							</tr>
						</tbody>
			   		</Table>
			    <Button className ="player_button" onClick={handleShow}>
			      View All Races
			    </Button>
			    <Modal show={show} onHide={handleClose}>
			     	<Modal.Header closeButton>
			       		<Modal.Title>All Races</Modal.Title>
			      	</Modal.Header>
			      	<Modal.Body>{allPlayerRaces.map(race =>{
			      	const parseISOString = (s) => moment(s).format('LLL')
			      	return(
			      		<div className="race_container">
				      		<p>{parseISOString(race.createdAt)}</p>
				      		<p>{race.standings.toString()}</p>
			      		</div>
			      		)
			      	})}
			      	</Modal.Body>
					<Modal.Footer>
						<Button variant="secondary" onClick={handleClose}>
					  		Close
						</Button>
					</Modal.Footer>
			    </Modal>
			    <p>
			      {player.name} finished first {player.results[1]} times. Second {player.results[2]} times, third {player.results[3]}
			        times..... and on {player.results[4]} occasions finished in dead last.
			    </p>
			    <p>Below is the graph to show the players ELO progress after each game</p>
	  		</Container>
		<Graph player={player.eloData.scorearray} />
	</Jumbotron>
		)
}

export default CenterPlayer;