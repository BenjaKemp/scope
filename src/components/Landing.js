import React from "react";
import Jumbotron from 'react-bootstrap/Jumbotron';
import Container from 'react-bootstrap/Container';


const Landing = ({players, races}) => {

	let individualBattles = players.reduce( (total, currentValue) => total + currentValue.eloData.scorearray.length, 0)
	let podium = players.slice(0,3);
	let numberOfMatches = races.length;

	return(
		<Jumbotron className ="container-full-bg landing_background" fluid>
		  <Container>
			  <h3 className="title">We've crunched the numbers on {numberOfMatches} seperate races</h3>
			  <h3 className="title">comprising {individualBattles} individual battles</h3>
			  <h3 className="title">The column on the left shows racer positions in ascending order</h3>
			  <ol className="podium">
			    <li>{podium[0].name}</li>
			    <li>{podium[1].name}</li>
			    <li>{podium[2].name}</li>
			  </ol>
		  </Container>
		</Jumbotron>
		)
}


export default Landing;