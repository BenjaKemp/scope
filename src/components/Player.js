import React from "react";
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';
import {Link, BrowserRouter} from "react-router-dom";

const Player  =  ({ player})  => {

	return(
		<Card className="card_wrapper">
		  <Card.Img variant="top" src={require(`../assets/${player.name.replace(/\s/g,'')}.png`)}  />
		  <Card.Body>
		    <Card.Title>{player.name}</Card.Title>
		    <BrowserRouter>
			    <Link exact to={`/playerdetails/${player.id}`}>
			    	<Button  className="player_button" variant="primary">Details</Button>
			    </Link>
			</BrowserRouter>
		  </Card.Body>
		</Card>
		)
}


export default Player;