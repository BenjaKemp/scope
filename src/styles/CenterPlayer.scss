.center_player_background {
	margin-top: 3px;
   	padding: 1;
    border-radius: 5px;
    border: black solid;
    height: 100%;

}

.race_container{
	border-radius: 5px;
	border: black solid;
	margin-bottom: 0.3rem;
}