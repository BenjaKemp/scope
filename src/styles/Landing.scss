
@import 'styles/_colors.scss'; 

body {
	text-align: center;
	font-family: Open Sans;
	background-color: $marioRed;
	margin: 0;
	padding: 0;
}

.landing_background {
	margin-top: 3px;
    padding-right: 0;
    padding-left: 0;
    border-radius: 5px;
    border: black solid;
    height: 100%;
}

.podium {
  display: flex;
  align-items: flex-end;
  padding: 0;
  margin: 5vmin;
  font-variant: small-caps;
}
.podium li {
  position: relative;
  list-style: none;
  background: linear-gradient(to right, black, 30%, silver 40%, 55%, black);
  flex: 1;
  padding: 1rem;
  margin: 0 0.5rem;
  box-sizing: border-box;
  text-align: center;
  font-size: calc(0.8rem + 2vw);
  text-shadow:
    -1px -1px black,
    +1px -1px black,
    +1px +1px black,
    -1px +1px black,
    -1px -1px 1px lightBlue,
    +1px -1px 1px lightBlue,
    +1px +1px 1px lightBlue,
    -1px +1px 1px lightBlue;
  border-top: solid 0.7rem;
  border-bottom: solid 0.2rem rgba(0,0,0,0.7);
  border-radius: 50% / 0.5rem 0.5rem 1rem 1rem;
}
.podium li::before {
  content: "";
  position: absolute;
  top: -1px;
  left: 0;
  width: 100%;
  border-bottom: solid 0.7rem;
  border-radius: 0 0 50% 50% / 0 0 1rem 1rem;
}
.podium li:first-child {
  color: #fc3;
  height: 45vh;
  min-height: 6em;
  order: 2;
}
.podium li:nth-child(2) {
  color: #ccd;
  height: 30vh;
  min-height: 4em;
  order: 1;
}
.podium li:nth-child(3) {
  color: #b62;
  height: 15vh;
  min-height: 2em;
  order: 3;
}

@media (max-width: 24em) {
  .podium {
    flex-direction: column;
    align-items: stretch;
  }
  .podium li:nth-child(n) {
    order: 0;
    height: 4em;
    min-height: 0;
    margin: 0.5rem 0;
  }
}