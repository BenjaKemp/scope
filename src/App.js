import React from 'react';
import backendApi from './utils/api'
import Player from './components/Player'
import Landing from './components/Landing'
import Dashboard from './components/Dashboard'
import NewGame from './components/NewGame'
import CenterPlayer from './components/CenterPlayer'
import Col from 'react-bootstrap/Col';
import Container from 'react-bootstrap/Container';
import Jumbotron from 'react-bootstrap/Jumbotron';
import Spinner from 'react-bootstrap/Spinner';
import Row from 'react-bootstrap/Row';
import { v4 as uuidv4 } from 'uuid';

import {
  BrowserRouter as Router,
  Switch,
  Route
} from "react-router-dom";
import 'bootstrap/dist/css/bootstrap.min.css';

const EloRating = require('elo-rating');
const io = require("socket.io-client");
const ioClient = io.connect("http://localhost:8000");

 

export default class App extends React.Component {
  constructor(props) {
    super(props);

    this.state = {};

    ioClient.on("updated", (msg) => {
       console.info(msg)
       this.refreshAllData()
    });
  }

  componentDidMount() {
    this.refreshAllData()
    
  }

  refreshAllData(){

    backendApi.callMatchesApi()
    .then(response => {
          let playerObject = {};
          let matches = response.list.map(obj=> ({ ...obj, id: uuidv4() })).sort(function(a,b){return new Date(a.updatedAt) - new Date(b.updatedAt)});
          matches.forEach(el => {
            el.standings.forEach((driver, i) => {
              let position = i +1;
              if(!playerObject[driver]){ 
                playerObject[driver] = {}
              }
              if(!playerObject[driver][position]){ 
                playerObject[driver][position] = 1
              }
              else{
                playerObject[driver][position] += 1
              }
            })
          })

      let keys = Object.keys(playerObject);

      let statePlayer = []

      keys.forEach(el => {
        let ELOViewsObject =  Object.assign(...keys.map(key => ({ [key]: {'ELO':1500, 'games':0, 'scorearray': []} })));

        for (var i = 0; i < matches.length; i++) {
          let gameOrders = matches[i].standings

          for (var j = 0; j < gameOrders.length; j++) {
            for (var k = j+1; k < gameOrders.length; k++) {
              let currentPlayer = gameOrders[j]
              let playerBehind = gameOrders[k]

              var regex = /[.,\s]/g;

              let objectkey = [currentPlayer,playerBehind].sort().toString()
                  objectkey = objectkey.replace(regex, '');

              if(ELOViewsObject[objectkey]){
                ELOViewsObject[objectkey] +=1
              }else{
                ELOViewsObject[objectkey] = 1
              }

              ELOViewsObject[currentPlayer]['games'] += 1
              ELOViewsObject[playerBehind]['games'] += 1 

              let Elo = EloRating.calculate(ELOViewsObject[currentPlayer]['ELO'] , ELOViewsObject[playerBehind]['ELO'], true, ELOViewsObject[currentPlayer]['games']);
     
              let winnerPoints = Elo.playerRating
              let loserPoints =  Elo.opponentRating
              

              ELOViewsObject[currentPlayer]['scorearray'].push(winnerPoints)    
              ELOViewsObject[playerBehind]['scorearray'].push(loserPoints)

              ELOViewsObject[currentPlayer]['pointsscored'] = winnerPoints - ELOViewsObject[currentPlayer]['ELO']
              ELOViewsObject[playerBehind]['pointsscored'] = loserPoints -ELOViewsObject[playerBehind]['ELO']

              ELOViewsObject[currentPlayer]['ELO'] = winnerPoints
              ELOViewsObject[playerBehind]['ELO'] = loserPoints
             
            }
          }
        }
        let playerMatches = matches.filter(match => match.standings.includes(el));

        statePlayer.push(
          { 
            id: uuidv4(),
            name: el,
            matches: playerMatches,
            eloData: ELOViewsObject[el],
            results: playerObject[el]
          })
      })

      statePlayer.sort((a, b) => b.eloData.ELO - a.eloData.ELO)
      this.setState({...this.state, matches, statePlayer})
    })

  }

  renderPlayers(){

    let keys = Object.keys(this.state.statePlayer)
    return (
      <div >
          {keys.map((el, i) =>  <Player player={this.state.statePlayer[el]} />)}
      </div>
      )
  }

  render() {

    if (!this.state.statePlayer) {
      return (
        <Container fluid>
          <Spinner animation="border" role="status">
           <span className="sr-only">Loading...</span>
          </Spinner>
        </Container>
      )
    }
    return (
      <Router>
            <Jumbotron className ="jumbotron_app">
            <h2 className="title">Welcome to MarioKart League!</h2>
            </Jumbotron>
            <Dashboard player={this.state.statePlayer}/>
              <Row className = "container_wrapper landing_background ">
                <Col className="race_rankings_column">{this.renderPlayers()}</Col>
                  <Switch>
                    <Route exact path="/">
                          <Col xs={9}><Landing races={this.state.matches} players={this.state.statePlayer}/></Col>
                    </Route>
                    <Route exact path="/newgame">
                          <Col xs={9}><NewGame players={this.state.statePlayer} /></Col>
                    </Route>
                    <Route exact path="/playerdetails/:id">
                          <Col xs={9}><CenterPlayer players={this.state.statePlayer} races={this.state.matches}/></Col>
                    </Route>
                </Switch>
              </Row>
      </Router>
    );
  }
}