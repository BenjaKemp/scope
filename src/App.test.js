import React from 'react';
import { render } from '@testing-library/react';
import App from './App';
import backendApi from './utils/api'
import  Graph  from './components/Graphs';
import  Landing  from './components/Landing';
import  Player  from './components/Player';
import  CenterPlayer  from './components/CenterPlayer';
import TestRenderer from "react-test-renderer";
import playerData from './mocks/player.json'
import races from './mocks/races.json'


const playerObject = playerData[0]


describe('call made to back end and data returned', () => {
  it('should load user data', () => {
    return backendApi.callMatchesApi()
    .then(data => {
      expect(data).toBeDefined()
    })
  })
})

describe('Player', () => {
  test('snapshot renders', () => {
    const component = TestRenderer.create(<Player player={playerObject} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});







