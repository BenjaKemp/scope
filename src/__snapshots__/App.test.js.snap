// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CenterPlayer snapshot renders 1`] = `
<div
  className="card_wrapper card"
>
  <img
    className="card-img-top"
    src="Drybones.png"
  />
  <div
    className="card-body"
  >
    <div
      className="card-title h5"
    >
      Dry bones
    </div>
    <a
      href="/playerdetails/1a8a2b59-9b94-43ee-8831-76cc15be81cc"
      onClick={[Function]}
    >
      <button
        className="player_button btn btn-primary"
        disabled={false}
        type="button"
      >
        Details
      </button>
    </a>
  </div>
</div>
`;

exports[`Counter snapshot renders 1`] = `
<div
  className="card_wrapper card"
>
  <img
    className="card-img-top"
    src="Drybones.png"
  />
  <div
    className="card-body"
  >
    <div
      className="card-title h5"
    >
      Dry bones
    </div>
    <a
      href="/playerdetails/1a8a2b59-9b94-43ee-8831-76cc15be81cc"
      onClick={[Function]}
    >
      <button
        className="player_button btn btn-primary"
        disabled={false}
        type="button"
      >
        Details
      </button>
    </a>
  </div>
</div>
`;

exports[`Player snapshot renders 1`] = `
<div
  className="card_wrapper card"
>
  <img
    className="card-img-top"
    src="Drybones.png"
  />
  <div
    className="card-body"
  >
    <div
      className="card-title h5"
    >
      Dry bones
    </div>
    <a
      href="/playerdetails/1a8a2b59-9b94-43ee-8831-76cc15be81cc"
      onClick={[Function]}
    >
      <button
        className="player_button btn btn-primary"
        disabled={false}
        type="button"
      >
        Details
      </button>
    </a>
  </div>
</div>
`;

exports[`Player snapshot renders 2`] = `
<div
  className="card_wrapper card"
>
  <img
    className="card-img-top"
    src="Drybones.png"
  />
  <div
    className="card-body"
  >
    <div
      className="card-title h5"
    >
      Dry bones
    </div>
    <a
      href="/playerdetails/1a8a2b59-9b94-43ee-8831-76cc15be81cc"
      onClick={[Function]}
    >
      <button
        className="player_button btn btn-primary"
        disabled={false}
        type="button"
      >
        Details
      </button>
    </a>
  </div>
</div>
`;
